### PR TITLE
Support per-list options and algorithms on the new room list

### PR DESCRIPTION
**Pull request by turt2live**
For https://github.com/vector-im/riot-web/issues/13635

**This should be reviewable commit-by-commit, which is recommended.** Each commit should have words on what it does, and some of them are unreasonably large (sorry).

Docs are enclosed, including a fancy flowchart that may or may not help.

Gif is a bit quick, but the idea is that it shows the list algorithm changing:
per-list-opts mp4

### CODE REVIEW

File: docs/room-list-store.md [hunk 1]
```diff
@@ -2,20 +2,31 @@
 
 It's so complicated it needs its own README.
 
+![](img/RoomListStore2.png)
```
Comment (jryans) [inline docs/room-list-store.md:5]: Wow, impressive, thanks for adding this! 😄

File: docs/room-list-store.md [hunk 2]
```diff
@@ -2,20 +2,31 @@
...
+algorithm to call, instead of having all the logic in the room list store itself.
 
-Behaviour of the room list takes the shape of determining what features the room list supports, as well
-as determining where and when to apply the sorting algorithm in a tag. The importance algorithm, which
-is described later in this doc, is an example of an algorithm which makes heavy behavioural changes
-to the room list.
 
 Tag sorting is effectively the comparator supplied to the list algorithm. This gives the list algorithm
-the power to decide when and how to apply the tag sorting, if at all.
+the power to decide when and how to apply the tag sorting, if at all. For example, The importance algorithm,
```
Comment (jryans) [inline docs/room-list-store.md:27]:
```suggestion
the power to decide when and how to apply the tag sorting, if at all. For example, the importance algorithm, 
```

File: src/stores/room-list/RoomListStore2.ts [hunk 3]
```diff
@@ -64,6 +60,7 @@ export class RoomListStore2 extends AsyncStore<ActionPayload> {
         this.checkEnabled();
         for (const settingName of this.watchedSettings) SettingsStore.monitorSetting(settingName, null);
         RoomViewStore.addListener(this.onRVSUpdate);
+        this.algorithm.on(LIST_UPDATED_EVENT, this.onAlgorithmListUpdated);
```
Comment (jryans) [inline src/stores/room-list/RoomListStore2.ts:63]: Should you remove the listener somewhere?

File: src/stores/room-list/algorithms/Algorithm.ts [hunk 4]
```diff
@@ -441,5 +495,48 @@ export abstract class Algorithm extends EventEmitter {
...
+            // TODO: Be smarter and splice rather than regen the planet.
+            // TODO: No-op if no change.
+            await this.setKnownRooms(this.rooms);
+            return true;
+        }
+
+        if (cause === RoomUpdateCause.NewRoom) {
+            // TODO: Be smarter and insert rather than regen the planet.
+            await this.setKnownRooms([room, ...this.rooms]);
+            return;
```
Comment (jryans) [inline src/stores/room-list/algorithms/Algorithm.ts:511]: Return value?

File: src/stores/room-list/algorithms/list-ordering/OrderingAlgorithm.ts [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+    /**
+     * Handle a room update. The Algorithm will only call this for causes which
+     * the list ordering algorithm can handle within the same tag. For example,
+     * tag changes will not be sent here.
+     * @param room The room where the update happened.
+     * @param cause The cause of the update.
+     * @returns True if the update requires the Algorithm to update the presentation layers.
+     */
+    // XXX: TODO: We assume this will only ever be a position update and NOT a NewRoom or RemoveRoom change!!
```
Comment (jryans) [inline src/stores/room-list/algorithms/list-ordering/OrderingAlgorithm.ts:70]: Should we throw an error to enforce that assumption?

File: src/stores/room-list/RoomListStore2.ts (hunk 3 above)
Comment (turt2live) [reply src/stores/room-list/RoomListStore2.ts:63]: One day, yes. We don't tear down the store at all though, and never change the field.

File: src/stores/room-list/algorithms/list-ordering/OrderingAlgorithm.ts (hunk 5 above)
Comment (turt2live) [reply src/stores/room-list/algorithms/list-ordering/OrderingAlgorithm.ts:70]: Possibly, though the "be smarter" todos should also take care of this.

**Pull request merged**